android/src/main/cpp/graphics/objects/Quad2dTessellatedOpenGl.cpp

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright (c) 2021 Ubique Innovation AG <https://www.ubique.ch>
+ *
+ *  This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ *  SPDX-License-Identifier: MPL-2.0
+ */
+
+#include "Quad2dTessellatedOpenGl.h"
+#include "TextureHolderInterface.h"
+#include "TextureFilterType.h"
+#include <cmath>
+
+Quad2dTessellatedOpenGl::Quad2dTessellatedOpenGl(const std::shared_ptr<::BaseShaderProgramOpenGl> &shader)
+    : shaderProgram(shader) {}
+
+bool Quad2dTessellatedOpenGl::isReady() { return ready && (!usesTextureCoords || textureHolder); }
+
+std::shared_ptr<GraphicsObjectInterface> Quad2dTessellatedOpenGl::asGraphicsObject() { return shared_from_this(); }
+
+std::shared_ptr<MaskingObjectInterface> Quad2dTessellatedOpenGl::asMaskingObject() { return shared_from_this(); }
+
+void Quad2dTessellatedOpenGl::clear() {
+    std::lock_guard<std::recursive_mutex> lock(dataMutex);
+    if (ready) {
+        removeGlBuffers();
+    }
+    if (textureCoordsReady) {
+        removeTextureCoordsGlBuffers();
+    }
+    if (textureHolder) {
+        removeTexture();
+    }
+    ready = false;
+}
+
+void Quad2dTessellatedOpenGl::setIsInverseMasked(bool inversed) { isMaskInversed = inversed; }
+
+void Quad2dTessellatedOpenGl::setFrame(const Quad3dD &frame, const RectD &textureCoordinates, const Vec3D &origin, bool is3d) {
+    std::lock_guard<std::recursive_mutex> lock(dataMutex);
+    ready = false;
+    this->frame = frame;
+    this->textureCoordinates = textureCoordinates;
+    this->quadOrigin = origin;
+    this->is3d = is3d;
+}
+
+void Quad2dTessellatedOpenGl::setSubdivisionFactor(int32_t factor) {
+    if (factor != subdivisionFactor) {
+        subdivisionFactor = factor;
+    }
+}
+
+void Quad2dTessellatedOpenGl::setMinMagFilter(TextureFilterType filterType) {
+    std::lock_guard<std::recursive_mutex> lock(dataMutex);
+    textureFilterType = filterType;
+}
+
+void Quad2dTessellatedOpenGl::setup(const std::shared_ptr<::RenderingContextInterface> &context) {
+    std::lock_guard<std::recursive_mutex> lock(dataMutex);
+    if (ready) {
+        return;
+    }
+
+    computeGeometry(false);
+
+    std::shared_ptr<OpenGlContext> openGlContext = std::static_pointer_cast<OpenGlContext>(context);
+    programName = shaderProgram->getProgramName();
+    program = openGlContext->getProgram(programName);
+    if (program == 0) {
+        shaderProgram->setupProgram(openGlContext);
+        program = openGlContext->getProgram(programName);
+    }
+
+    prepareGlData(program);
+    prepareTextureCoordsGlData(program);
+
+    ready = true;
+}
+
+void Quad2dTessellatedOpenGl::computeGeometry(bool texCoordsOnly) {
+    // Data mutex covered by caller Quad2dTessellatedOpenGL::setup()
+    if (!texCoordsOnly) {
```

Comment:
Style: split function (tex-coords part and geometry part).
Style 2: clearer as function, take input as param and return result, assign at call site `vertices = computeVertexPositions(frame)`